Remove product scheduling from woocommerce_save_product_variation hook and prevent publish button spam

chickenn00dle opened the pull request:
Fixes #3192 

#### Changes proposed in this Pull Request

This removes the scheduling callback from the `woocommerce_save_product_variation` hook to avoid duplicate products in Stripe when publish is selected multiple times.

As far as I can tell, it should no longer be necessary to trigger this on both `save_post` and `woocommerce_save_product_variation` hooks since we were doing this to get around a price update issue and we are now creating price data on the fly when creating new subscriptions.


#### Testing instructions

1. Create a new variable subscription product, with multiple variations
2. Click the "Publish" button in quick succession multiple times
3. On `develop` you will see multiple instances of the same variation in Stripe, equal to the number of clicks from step 2
4. On this branch there should only be one instance of each variation in Stripe
5. Try to update the product name.
6. Ensure the update is reflected correctly in Stripe

- [x] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

chickenn00dle committed:
Remove product create/update on variation hook

This removes the scheduling callback from the
woocommerce_save_product_variation hook to avoid duplicate products in
Stripe when publish is selected multiple times.

james-allan reviewed (commented):
I think removing this hook is a good idea given the reasons you mentioned on the PR. ie It's no longer necessary given we don't need to update variations when they are saved independently of the product (for prices etc), however, this PR doesn't seem to prevent the duplicate product issue for me. 

Hitting the publish button multiple times still results in multiple products. **Note:** it only seems to affect publishing new products, hitting the update button for existing products is fine. At least in my testing.

From what I can tell, the issue occurs because when publishing a product, hitting the publish button multiple times results in multiple requests. 

You can tell by the URL that multiple request are being sent as the URL grows each time you click the button. 

https://user-images.githubusercontent.com/8490476/145948560-7700a296-4fce-49ca-950b-7a7db91970d7.mov

As far as I can tell, because these multiple requests get fired off in multiple individual requests, in parallel, the only option might be to store something when we send the request to prevent the other requests from processing it too. 🤷 

I'm not really sure.

chickenn00dle commented:
Thanks for looking into this @james-allan!

> Hitting the publish button multiple times still results in multiple products. Note: it only seems to affect publishing new products, hitting the update button for existing products is fine. At least in my testing.

Ah. It looks like this isn't the case on Firefox: https://cloudup.com/i4mBeOF6_Dz

Looking at the requests using dev tools, it looks like FF only actually processes the first one when the publish button is spammed. I checked Chrome just now, and it looks like they all end up being sent 😮‍💨

> As far as I can tell, because these multiple requests get fired off in multiple individual requests, in parallel, the only option might be to store something when we send the request to prevent the other requests from processing it too.

I wonder if instead of relying on the backend to prevent duplicates, we can just disable the publish button after the first click on the frontend?

Let me know what your thoughts on this are, then I'll go ahead and take a second pass at this issue.

james-allan commented:
>  I checked Chrome just now, and it looks like they all end up being sent

Oh nice find! 

>  we can just disable the publish button after the first click on the frontend?

That's an interesting idea. Looking into it a little more, WordPress core has some JS that appends to the current URL when the publish button is pressed when the post's status is auto-draft ([here](https://github.com/WordPress/WordPress/blob/4.8/wp-admin/js/post.js#L1203-L1214)). That code must have an issue in Chrome. I think limiting the button to 1 click is a decent idea and would be pretty low impact compared to storing data etc.

chickenn00dle committed:
Adds a script for product edit page

This adds a JS script for the product edit page that prevents duplicate
wcpay subs from being created when the submit button is spammed. This is
done by disabling the submit button after the first click for
subscription and variable subscription products.

chickenn00dle commented:
> I think limiting the button to 1 click is a decent idea and would be pretty low impact compared to storing data etc.

Great! I went with this approach in https://github.com/Automattic/woocommerce-payments/pull/3509/commits/783d78e4ce3eb2ac00af1688c89d3cc9dfc108ce

chickenn00dle committed:
Update changelogs

haszari commented:
FYI @shendy-a8c - this looks close, might be ready in time for WCPay 3.5 🙂

aprea commented on includes/subscriptions/assets/js/product-page.js:
```diff
@@ -0,0 +1,21 @@
+			$( '#publish' ).on( 'click', function ( e ) {
+				if (
+					$( 'select#product-type option:selected' )
+						.text()
```
We might want to use `.val()` here instead, using `.text()` might not work with l10n.

e.g. https://github.com/Automattic/woocommerce-subscriptions-core/blob/1.3.0/assets/js/admin/admin.js#L20

shendy-a8c replied:
Done in 6b18449.

shendy-a8c committed:
Merge branch 'develop' into fix/issue_3192

# Conflicts:
#	changelog.txt
#	readme.txt

shendy-a8c committed:
Use val() instead of text() to get product type.

shendy-a8c committed:
Merge branch 'develop' into fix/issue_3192

# Conflicts:
#	changelog.txt
#	readme.txt

shendy-a8c commented:
Can anyone produce the problem on `develop`? I have tested locally and using JN but somehow I can't make it so Stripe products are created in duplication. I have tried to click so fast. Maybe not fast enough?

However, on the PR branch, I see that as soon as I click the publish button, there is spinning animation, disabling the button. I have tested the normal flow to make sure the product can be purchased with no problem.

Changes looks good too.

haszari commented on changelog.txt:
```diff
@@ -15,6 +15,7 @@
 * Fix - Incorrect text when filtering subscriptions to no results.
 * Changed - Subscription products must have a recurring amount greater than $0.
 * Fix - Return correct product prices datatype in WCPay.
+* Fix - Prevent duplicate subscriptions from being created when submit button is spammed.
```
Suggest using clearer language here - `spammed` is colloquial, casual. E.g. `Prevent duplicate subscriptions from being created when submit button is repeatedly clicked` or similar.

shendy-a8c replied:
Updated in c173a4c.

haszari commented on includes/subscriptions/assets/js/product-page.js:
```diff
@@ -0,0 +1,21 @@
+						.val()
+						.includes( 'subscription' )
+				) {
+					$( this ).attr( 'disabled', true );
```
This feels like a pretty heavy workaround - adding a new jQuery JS file just to prevent double-submit (in a single browser). 

- Should we log this issue with WooCommerce Core?
- Could this double-submit issue affect non-subscription products too? If not, why are subscription products unique?

shendy-a8c replied:
> Should we log this issue with WooCommerce Core?

That is a good question. I agree that preventing multiple clicks to a submit button should be a standard across and not just desirable for subscription specific product.

haszari commented:
> Can anyone produce the problem on develop? I have tested locally (with WC 5.6.0) and using JN (with WC 6.0.0) but somehow I can't make it so Stripe products are created in duplication. I have tried to click so fast. Maybe not fast enough?

@a-danae found the issue originally, not sure if it was on local dev or elsewhere. Any tips on how to reproduce reliably @a-danae ?

@shendy-a8c / reviewers are you comfortable getting this into next release on Raz' behalf since he's afk? If there's any risk feel free to leave it for now.

I added some review comments - I don't see any major risk here, but the JS workaround doesn't seem to be adding a lot of value.

shendy-a8c committed:
Change the changelog wording.

shendy-a8c commented:
IMO because:
- This fix is low impact, that is duplicated Stripe products do not have a significant bad implication and rarely user clicks the publish button multiple times, and
- There is some discussion above (whether this fix should be on woo core) to resolve.

I am taking this PR of the [3.5 milestone](https://github.com/Automattic/woocommerce-payments/milestone/23).

If anyone think this needs to be shipped with 3.5, let me know.

haszari commented:
> I am taking this PR off the 3.5 milestone.

Thanks @shendy-a8c ! 

Now we need to figure out if we want to ship this in 3.6 or if we need to shelve this approach (or tweak it). Let's discuss in team call.

shendy-a8c commented:
Per pdjTHR-sb-p2#comment-620, closing this PR. @james-allan would create a new PR taking some of the changes in this PR, ie removing the `add_action()` and `remove_action()` for `woocommerce_save_product_variation`.

james-allan commented:
>  @james-allan would create a new PR taking some of the changes in this PR, ie removing the add_action() and remove_action() for woocommerce_save_product_variation.

While writing up the new PR for those changes I found there is a scenario where those hooks are still necessary. 

Without those hooks:

1. Create a variable product with a few variations. 
2. Publish that product. 
3. Notice in Stripe that all the variations are created.
4. Now add a new variation and hit the **'Save variations'** button. Do not press the **'Update'** post button. 
5. Without those hooks, the new variation will be created in WC but not at Stripe.

With that in mind, there's no need to remove them. I wasn't able to find a scenario where these hooks were harmful as there is already protection to avoid multiple calls leading to multiple saves in same request.

As I mentioned in the team meeting yesterday, if these hooks do become a problem we might need some other method for avoiding multiple requests via a request lock via a transient or similar.
